code/modules/crafting/craft.dm

```diff
@@ -0,0 +1,341 @@
+/datum/personal_crafting
+	var/busy
+	var/viewing_category = 1 //typical powergamer starting on the Weapons tab
+	var/list/categories = list(CAT_WEAPON,CAT_AMMO,CAT_ROBOT,CAT_FOOD,CAT_MISC,CAT_PRIMAL)
+	//var/datum/action/innate/crafting/button // Holdover from when crafting was still an action button. TODO: Remove this before I commit.
+
+
+
+
+
+/*	This is what procs do:
+	get_environment - gets a list of things accessable for crafting by user
+	get_surroundings - takes a list of things and makes a list of key-types to values-amounts of said type in the list
+	check_contents - takes a recipe and a key-type list and checks if said recipe can be done with available stuff
+	check_tools - takes recipe, a key-type list, and a user and checks if there are enough tools to do the stuff, checks bugs one level deep
+	construct_item - takes a recipe and a user, call all the checking procs, calls do_after, checks all the things again, calls del_reqs, creates result, calls CheckParts of said result with argument being list returned by deel_reqs
+	del_reqs - takes recipe and a user, loops over the recipes reqs var and tries to find everything in the list make by get_environment and delete it/add to parts list, then returns the said list
+*/
+
+
+
+
+/datum/personal_crafting/proc/check_contents(datum/crafting_recipe/R, list/contents)
+	main_loop:
```

Comment:
Knew somebody would try to pull that "It's goto!" bullshit, we've had this discussion before at /tg/, it's nowhere near as abusive as ```goto```, and is perfectly fine to use.

And just as a reminder, computers would be nothing without ```goto```, loops and other constructs compile down to them, you can't avoid them, they're only wrong when abused, and it's incredibly difficult to abuse loop labels (I can't even think of a scenario with a loop label that is considered "abuse", overuse isn't even a problem as they're better than temp vars and extraneous ifs that cause branch prediction to fail, in fact, the only time I wouldn't advocate a loop label is when a loop isn't nested, in which case ```continue```/```break``` without a label is the best option)

Also more procs Vamp? really? DM one of the worst things about DM is it's proc call overhead, it's why recursion isn't viable in DM, and it's why it's not a valid approach for it here either.

tl;dr stop.